File: aspnet/Tide.Asgard.AspNetCore/Tide.Asgard.Scheduler.Postgres/PostgresNotifier.cs
// Copyright (c) Tide Foundation Limited. All rights reserved.
// Licensed under the Tide Community Open Code License. See LICENSE in the project root.

using Npgsql;
using Tide.Asgard.Scheduler.Execution;

namespace Tide.Asgard.Scheduler.Postgres;

// Wakes workers across processes. NOTIFY goes out through the data source, LISTEN
// sits on a connection of its own, because notifications are delivered to the
// session that issued the LISTEN and a pooled connection would deliver them to
// whichever caller happened to get it next.
public sealed class PostgresNotifier : IJobNotifier, IAsyncDisposable
{
	public const string JobChannel = "asgard_jobs";

	private readonly NpgsqlDataSource _dataSource;
	private readonly SemaphoreSlim _listenerGate = new(1, 1);
	private NpgsqlConnection? _listener;

	public PostgresNotifier(NpgsqlDataSource dataSource) => _dataSource = dataSource;

	public async Task NotifyAsync(CancellationToken ct = default)
	{
		try
		{
			await using var command = _dataSource.CreateCommand($"select pg_notify('{JobChannel}', '')");
			await command.ExecuteNonQueryAsync(ct);
		}
		catch (Exception) when (!ct.IsCancellationRequested)
		{
			// Deliberately swallowed. A worker that cannot announce new work
			// still enqueued it, and the next poll finds it. Latency, not
			// correctness.
		}
	}

	public async Task WaitAsync(long timeoutMs, CancellationToken ct = default)
	{
		if (ct.IsCancellationRequested) return;

		using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
		timeout.CancelAfter(TimeSpan.FromMilliseconds(timeoutMs));

		try
		{
			var listener = await ListenerAsync(ct);

			// Returns as soon as a notification arrives on any listened channel,
			// or when the linked token fires at the timeout.
			await listener.WaitAsync(timeout.Token);
		}
		catch (OperationCanceledException)
		{
			// Timed out or shutting down. Both are normal.
		}
		catch (Exception)
		{
			// A dropped connection means notifications stop arriving. Discard it
			// so the next wait reconnects, and fall back to the timeout, which is
			// what a worker without a notifier does anyway.
			await DiscardListenerAsync();
			await Task.Delay(TimeSpan.FromMilliseconds(timeoutMs), CancellationToken.None);
		}
	}

	private async Task<NpgsqlConnection> ListenerAsync(CancellationToken ct)
	{
		await _listenerGate.WaitAsync(ct);
		try
		{
			if (_listener is { State: System.Data.ConnectionState.Open }) return _listener;

			_listener = await _dataSource.OpenConnectionAsync(ct);
			await using var command = new NpgsqlCommand($"listen {JobChannel}", _listener);
			await command.ExecuteNonQueryAsync(ct);
			return _listener;
		}
		finally
		{
			_listenerGate.Release();
		}
	}

	private async Task DiscardListenerAsync()
	{
		await _listenerGate.WaitAsync();
		try
		{
			if (_listener is not null) await _listener.DisposeAsync();
			_listener = null;
		}
		finally
		{
			_listenerGate.Release();
		}
	}

	public async ValueTask DisposeAsync()
	{
		await DiscardListenerAsync();
		_listenerGate.Dispose();
	}
}

File: src/scheduler/execution/PostgresNotifier.ts
// Copyright (c) Tide Foundation Limited. All rights reserved.
// Licensed under the Tide Community Open Code License. See LICENSE in the project root.

import { JobNotifier } from "./JobNotifier";
import { SqlClient } from "./PostgresJobStore";

export const JOB_CHANNEL = "asgard_jobs";

// LISTEN needs a connection of its own: notifications are delivered to the
// session that issued the LISTEN, so a pooled connection would deliver them to
// whichever caller happened to get that connection next. A node-postgres Client
// satisfies this as is, which is why nothing here imports pg.
export interface ListenClient {
    query(text: string): Promise<unknown>;
    on(event: string, listener: (...args: any[]) => void): unknown;
    removeListener?(event: string, listener: (...args: any[]) => void): unknown;
}

// Wakes workers across processes. NOTIFY goes out through the pool, LISTEN sits
// on a dedicated connection the host supplies already connected:
//
//   const listener = new Client({ connectionString });
//   await listener.connect();
//   const notifier = new PostgresNotifier(pool, listener);
export class PostgresNotifier implements JobNotifier {
    private listening = false;
    private waiters: Array<() => void> = [];

    constructor(
        private readonly sql: SqlClient,
        private readonly listener: ListenClient
    ) { }

    async notify(): Promise<void> {
        // Deliberately swallowed. A worker that cannot announce new work still
        // enqueued it, and the next poll finds it.
        try {
            await this.sql.query(`select pg_notify('${JOB_CHANNEL}', '')`);
        } catch {
            // Latency, not correctness.
        }
    }

    async wait(timeoutMs: number, signal?: AbortSignal): Promise<void> {
        if (signal?.aborted) return;

        try {
            await this.ensureListening();
        } catch {
            // Fall back to waiting out the timeout, which is what a worker
            // without a notifier does anyway.
        }

        return new Promise<void>(resolve => {
            let settled = false;

            const finish = () => {
                if (settled) return;
                settled = true;
                clearTimeout(timer);
                signal?.removeEventListener("abort", finish);
                this.waiters = this.waiters.filter(w => w !== finish);
                resolve();
            };

            const timer = setTimeout(finish, timeoutMs);
            signal?.addEventListener("abort", finish);
            this.waiters.push(finish);
        });
    }

    private async ensureListening(): Promise<void> {
        if (this.listening) return;
        this.listening = true;

        this.listener.on("notification", (message: { channel?: string }) => {
            if (message?.channel !== JOB_CHANNEL) return;
            this.wake();
        });

        // A dropped connection means notifications stop arriving. Waking
        // everyone and re-listening next time keeps that to a latency problem.
        this.listener.on("error", () => {
            this.listening = false;
            this.wake();
        });

        await this.listener.query(`listen ${JOB_CHANNEL}`);
    }

    private wake(): void {
        const woken = this.waiters;
        this.waiters = [];
        for (const resolve of woken) resolve();
    }
}
